text/3857-cfg-version.md

```diff
@@ -0,0 +1,648 @@
+
+Rust does appear in some parts of the language,
+but is capitalized like with [`repr(Rust)`](https://doc.rust-lang.org/reference/type-layout.html?#the-rust-representation).
+However, the convention for `--cfg`s is generally lower case.
+
+### Pre-release
+
+When translating `rustc --version` to a language version, we have several choices when it comes to pre-releases, including:
+- Treat the nightly as fully implementing that language version
+- Treat the nightly as not implementing that language version at all, only the previous
+- Leave a marker that that language version is incomplete, while the previous language version is complete
+
+In RFC 2523, this was left as an
+[unresolved question](https://rust-lang.github.io/rfcs/2523-cfg-path-version.html#unresolved-questions).
+
+The initial implementation treated nightlies as complete.
+This was [changed to incomplete](https://github.com/rust-lang/rust/pull/72001) after
+[some discussion](https://github.com/rust-lang/rust/issues/64796#issuecomment-624673206).
+In particular, this is important for
+- the case of package `bleeding-edge` starting to use a new feature behind `#[cfg(since)]` and package `nightly-only` has their toolchain pinned to a nightly before the feature was stabilized (to ensure consistent behavior of unstable features), package `nightly-only` cannot add or update their dependency on `bleeding-edge` without getting a "feature gate needed" error.
+- bisecting nightlies.
+
+This was [changed back to complete](https://github.com/rust-lang/rust/pull/81468) after
+[some more discussion](https://github.com/rust-lang/rust/issues/64796#issuecomment-634546711).
+In particular, this is important for
+- keeping friction down for packages preparing for stabilized-on-nightly features as their `#[cfg(since)]`s can be inserted and "just work" which can be important for getting feedback quickly while the feature is easier to adapt to feedback that can be gained from these users
+  - releasing the package while its in this state puts it at risk to be broken if the feature is changed after stabilization
+
+For RFC 2523, they settled on pre-releases being incomplete,
+favoring maintainers to adopt stabilized-on-nightly features immediately
+while letting people on pinned nightlies or bisecting nightlies to set a `-Z` to mark the version as incomplete.
+
+In this RFC, we settled on translating `-nightly` to `-incomplete` because:
+- Maintainers can adopt stabilized-on-nightly features with `#[cfg(since(rust, "1.100.0-0"))]` (the lowest pre-release for `1.100.0`), keeping friction low while explicitly acknowledging that the unstable feature may change
+- Allows build scripts to experiment with other logic when approximating the vendor version from the language version with less of a chance of needing to invoke `rustc` (e.g. detecting nightly)
+- It provides extra context when approximating the vendor version from the language version when populating build information
```

Comment:
@dtolnay considering you were a major driver for where things landed for the previous work on how `rust_version` treated nightlies, I'd be interested in your thoughts on this alternative take that this new RFC has.